bluesky/utils.py

```diff
@@ -1006,30 +1007,34 @@ def update(self, pos, *,
         self.draw()
 
     def draw(self):
-        if (time.time() - self.creation_time) < self.delay_draw:
-            return
-        self.drawn = True
-        for meter in self.meters:
-            tqdm.status_printer(self.fp)(meter)
-            self.fp.write('\n')
-        self.fp.write(_unicode(_term_move_up() * len(self.meters)))
+        with threading.Lock():
+            if (time.time() - self.creation_time) < self.delay_draw:
+                return
+            if self.done:
+                return
+            for meter in self.meters:
+                tqdm.status_printer(self.fp)(meter)
+                self.fp.write('\n')
+            self.fp.write(_unicode(_term_move_up() * len(self.meters)))
+            self.drawn = True
 
     def _ensure_draw(self):
         # Ensure that the progress bar is drawn at least once after the delay.
         time.sleep(self.delay_draw)
-        if (not self.done) and (not self.drawn):
-            self.draw()
+        with threading.Lock():
+            if (not self.done) and (not self.drawn):
+                self.draw()
 
     def clear(self):
-        self.done = True
-        if self.drawn:
-            for meter in self.meters:
-                self.fp.write('\r')
-                self.fp.write(' ' * self.ncols)
-                self.fp.write('\r')
-                self.fp.write('\n')
-        self.fp.write(_unicode(_term_move_up() * len(self.meters)))
-        self.meters.clear()
+        with threading.Lock():
+            self.done = True
+            if self.drawn:
+                for meter in self.meters:
+                    self.fp.write('\r')
+                    self.fp.write(' ' * self.ncols)
+                    self.fp.write('\r')
+                    self.fp.write('\n')
+                self.fp.write(_unicode(_term_move_up() * len(self.meters)))
```

Comment:
The indentation change here is the crucial change, but the additional locking doesn't hurt.